auth: Add apple auth support.

chdinesh1089 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Add support for Sign In with Apple.

**Testing Plan:** <!-- How have you tested? -->
Added test class for apple auth.

closes: #14168

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# (1) Visit https://developer.apple.com/account/resources/,
+# Follow the instructions in
+# https://help.apple.com/developer-account/?lang=en#/dev77c875b7e
+# to create a Sign In Certificate.
```
Let's be sure to consistently use the same terminology as Apple's docs do. The thing created through these instructions is a "private key", or a "Sign in with Apple private key", not a "certificate".

(This particular distinction is actually a very important one, beyond matching Apple's terminology. Certificates generally are not secret -- they are public things that you might share widely, showing them to anyone you want to talk to. Private keys are, well, private -- it's essential not to let anyone else see a copy of one.)

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# to Create a Services ID. In services ID, fill the redirect URI with
+# a value like `http://zulip.example.com/complete/apple-id` .
+#
+# (3) You'll now have a client_id(eg: com.application.your), aka "Service ID",
```
similarly, this is a "Services ID", I think

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# a value like `http://zulip.example.com/complete/apple-id` .
+#
+# (3) You'll now have a client_id(eg: com.application.your), aka "Service ID",
+# a Key ID and a Private key file. Use the client_id as SOCIAL_AUTH_APPLE_ID_CLIENT
```
and I see "key identifier" in the docs, but not "Key ID"

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# a value like `http://zulip.example.com/complete/apple-id` .
+#
+# (3) You'll now have a client_id(eg: com.application.your), aka "Service ID",
+# a Key ID and a Private key file. Use the client_id as SOCIAL_AUTH_APPLE_ID_CLIENT
```
And is there something in particular that the name "client_id" refers to, or can we just say "Services ID" here again?

chdinesh1089 replied:
we can just say "service ID" but the `python-social-auth` has it named `SOCIAL_AUTH_APPLE_ID_CLIENT`. so thought using `client_id` is good.
I referred the doc in this [PR](https://github.com/python-social-auth/social-docs/pull/68/files#diff-956fdef252a7f17b6e232fe3c182ce05) for writing those instructions, which is why I used the name `client_id`. I think changing it and having just "service ID" is fine as we'll have the instruction clearly to set that `SOCIAL_AUTH_APPLE_ID_CLIENT`.

gnprice replied:
Sounds good. Note though that it's "Services ID" -- Apple's docs consistently spell and capitalize it that way, and to avoid confusion we should do the same.

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# (4) Put the Private key file in zulip server in
+# /etc/zulip/apple/zulip-private-key.key. Make sure to set proper permissions
+# with the following commands
+# chown -R zulip.zulip /etc/zulip/apple/
```
Let's use a colon in the argument to `chown`, not a dot -- so write `zulip:zulip`, not `zulip.zulip`.

TIL that a dot seems to work here! But that's undocumented (in at least the manpage I get with `man chown`, for GNU `chown`), and I suspect nonstandard. The conventional thing is a colon.

gnprice commented on docs/development/authentication.md:
````diff
@@ -42,6 +42,25 @@ details worth understanding:
+  ```
+  chown -R zulip.zulip /etc/zulip/apple/
+  chmod 640 /etc/zulip/apple/zulip-private-key.key
+  ```
````
Same comments apply as to the other, prod version of these instructions.

gnprice commented on zproject/backends.py:
```diff
@@ -1392,6 +1396,13 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+    sort_order = 200
+    name = "apple-id"
+    auth_backend_name = "Apple"
+    display_icon = "/static/images/landing-page/logos/apple-icon.png"
```
One other comment:

Does this produce a button that follows the guidelines Apple provides specifically for "Sign in with Apple" buttons? See here:
https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/

There is provision there for making a custom button. But there are a number of details if so.

gnprice replied:
Alternatively, Apple provides on a CDN complete buttons in a range of styles:
https://developer.apple.com/documentation/signinwithapplejs/incorporating_sign_in_with_apple_into_other_platforms

Note in particular the `locale` parameter, for controlling the language of the text.

chdinesh1089 replied:
> Does this produce a button that follows the guidelines Apple provides specifically for "Sign in with Apple" buttons?

Didn't know much about `locale` parameter, So thought we'll have problem with translations.
will try to using CDN complete buttons provided by apple.

chdinesh1089 replied:
The buttons provided by apple have the logo centre aligned meaning it was just infront of "Sign in with Apple" text. If we use that it makes that look very odd from other "Log in with X" buttons. May be should create a custom one with their guidelines. If we are doing it we might want to adjust other buttons also( like changing "Log in" to "Sign in", maintaining the ratio of font-size to button size suggested by apple), so this would blend in the page well like the other buttons.

gnprice replied:
Hmm, I see.

I'd like to avoid blocking this on any broader design changes to our auth buttons. So for getting this merged, what I'd like to prioritize is
* Have the Apple button compatible with Apple's guidelines.
* Keep the work simple, to focus on the rest of it and getting it to something we can merge and deploy.

I think it's OK to push to a followup, for after this is merged, the task of:
* Make the design of the Apple auth button in line with our other auth buttons.

How's that sound?

For that set of priorities, I think the simplest way will most likely be to use a button from the CDN, and accept that the logo is placed differently from the other buttons' logos. But a custom button following the guidelines would also be good if that's easier.

chdinesh1089 replied:
Okay, for now, we'll use the button from CDN.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -455,8 +455,11 @@ def start_social_login(request: HttpRequest, backend: str, extra_arg: Optional[s
         extra_url_params = {'idp': extra_arg}
 
+        if backend == "apple-id":
+            backend = "apple_id"
```
Why do we need this conversion?

mateuszmandera replied:
It would be best if we can avoid this kind of thing

chdinesh1089 replied:
It's because we are doing `backend.upper()` below to get `key_setting`. 
should I change it and add a new line with `if backend == "apple-id"`?

gnprice replied:
How about we make the backend name just `"apple"`, and the setting names just `SOCIAL_AUTH_APPLE_TEAM` and so on?

That would resolve the need for any wrinkle in this logic.

It'd also make the settings names rather easier to understand, I think -- it's kind of confusing that e.g. the setting for the "Team ID" has a name that ends in `ID_TEAM`. Which is happening because of this collision, or coincidence, that the word "ID" is being used for two different things there in quick succession.

I see that upstream sets `name = "apple-id"`, but we're subclassing it and overriding that anyway already.

gnprice replied:
Alternatively, "appleid". That's what Apple uses in the URLs that appear in this system -- like https://appleid.apple.com/auth/authorize and https://appleid.cdn-apple.com/appleid/button . And it avoids the confusion of reading e.g. "APPLE_ID_TEAM" like "Apple / ID team", as well as any need to juggle hyphens vs. underscores.

mateuszmandera commented on zproject/default_settings.py:
```diff
@@ -47,6 +47,11 @@
 
 # Social auth; we support providing values for some of these
 # settings in zulip-secrets.conf instead of settings.py in development.
+SOCIAL_AUTH_APPLE_ID_CLIENT = get_secret('social_auth_apple_id_client')
```
What's the reason for using `get_secret` for  the id_client?

chdinesh1089 replied:
I'm thinking we'll have id_client in `dev_secrets.conf` or `zulip_secrets.conf`.

gnprice replied:
(This is another example of where it's confusing that there's an extra "id" here before "client" :smile: -- the name is trying to say "social auth / Apple ID / client", but as you can see, it's very natural to read it instead as "social auth / Apple / id_client". Even though "id_client" is a bit of a mysterious name in itself.)

chdinesh1089 replied:
I think using `get_secret` should be right here, because during development to this when we want to set this value to something, it becomes an obstacle while pushing to github. In production we ask the user to set this value in `prod_settings_template.py`.

mateuszmandera commented:
Posted two minor comments, but the code looks right so far. I'm expecting that it'll turn out that the upstream implementation of the backend will have some issues pop up or we'll need some adjustments, so some overriding of the class methods of the backend will be needed like with other backends. That'll be figured out when we have valid keys so that we can test this though - and perhaps I should take care of that part, since I think I'll be able to find time for soon and I already did a bunch of digging into the internals of how this authentication method works?

Other thing that needs to be done (and isn't blocked due to us not having valid certs from Apple yet) iis the configuration for tests to get them working. For that we'll want dummy values, that are in the correct format (so the libraries accept and process them correctly), but generated by us, not Apple.

 SAML is done this way (in test_settings.py):
```
SOCIAL_AUTH_SAML_SP_PUBLIC_CERT  = get_from_file_if_exists("zerver/tests/fixtures/saml/zulip.crt")
SOCIAL_AUTH_SAML_SP_PRIVATE_KEY = get_from_file_if_exists("zerver/tests/fixtures/saml/zulip.key")
```
(we have self-generated (by me) cert and key in those files). Can you look into this too?

chdinesh1089 commented:
> Can you look into this too?

I'll try generating the files.

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# chmod 640 /etc/zulip/apple/zulip-private-key.key
+#
+#SOCIAL_AUTH_APPLE_ID_CLIENT = <your Services ID>
+#SOCIAL_AUTH_APPLE_ID_KEY = <your Key ID>
```
Do you want a key *identifier* here, or an actual key?

Some of the other code in this PR looks like it wants to interpret this as an actual key. (Or in other words: a secret.) But these instructions say to put the mere key identifier here, which is just a kind of name for a key.

If the instructions are right, the name should agree with it. If not, vice versa :slightly_smiling_face:

gnprice replied:
... Oof, I see, the confusion is there in the [upstream code we're using](https://github.com/python-social-auth/social-core/pull/409/files):
```python
    def generate_client_secret(self):
￼        now = int(time.time())
￼        client_id = self.setting("CLIENT")
￼        team_id = self.setting("TEAM")
￼        key_id = self.setting("KEY")
￼        private_key = self.get_private_key()
```

So be it, then.

mateuszmandera commented on templates/zerver/login.html:
```diff
@@ -113,7 +113,8 @@ <h1 class="get-started">{{ _("Log in to Zulip") }}</h1>
                             <button id="login_{{ backend.button_id_suffix }}" class="login-social-button"
-                              {% if backend.display_icon %} style="background-image:url({{ backend.display_icon }})"  {% endif %}>
+                              {% if backend.display_name == "Apple" %} style="background-image: url(https://appleid.cdn-apple.com/appleid/button/?width=328&height=46&color=white&locale={{locale}});"{% endif %}>
+                                {% if backend.display_icon %} style="background-image:url({{ backend.display_icon }})"  {% endif %}>
```
Shouldnt this be `{% elif backend.display_icon %}`? Could you also show a screenshot of the button?

chdinesh1089 replied:
ah, I messed it up because I was having linting issues as `<button` closing tag`>` wasn't set properly. Now I've edited it, but had to set `font-size: 0px;` in css for apple button.

mateuszmandera commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# (2) Follow instructions in
+# https://help.apple.com/developer-account/?lang=en#/dev1c0e25352.
+# to Create a Services ID. In services ID, fill the redirect URI field
+# with a value like `http://zulip.example.com/complete/apple-id`.
```
`/apple` now that we changed backend name

mateuszmandera commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,43 @@
+# (3) You'll now have a "Service ID"(eg: com.application.your),a "Key Identifier"
+# and a Private key file. Use the "Service ID" as `SOCIAL_AUTH_APPLE_ID_CLIENT`
+# and "Key Identifier" as `SOCIAL_AUTH_APPLE_ID_KEY` and if you have a "Team ID"
+# that can be set as `SOCIAL_AUTH_APPLE_ID_TEAM` here.
```
we dont have the `_ID_` part anymore

mateuszmandera commented on zproject/settings.py:
```diff
@@ -979,6 +979,7 @@ def zulip_path(path: str) -> str:
                                         'multiuse_object_key']
 SOCIAL_AUTH_LOGIN_ERROR_URL = '/login/'
 
+SOCIAL_AUTH_APPLE_SECRET = get_from_file_if_exists("etc/zulip/apple/zulip-private-key.key")
```
`/etc`/, typo :slightly_smiling_face:

mateuszmandera commented on templates/zerver/apple-error.md:
```diff
@@ -0,0 +1,14 @@
+  in `/etc/zulip/apple/zulip-private-key.key` in zulip server
+  with proper permissions set.
+
+* Navigate back to the login page and attemp Apple auth flow again.
```
attempt - typo :slightly_smiling_face:

mateuszmandera replied:
"attempt Sign in with Apple flow" would perhaps be a nicer phrasing?

mateuszmandera commented on templates/zerver/apple-error.md:
```diff
@@ -0,0 +1,14 @@
+  enable "Sign In with Apple" for an app at
+  [Certificates, Identifiers & Profiles](https://developer.apple.com/account/resources/).
+
+* You have set `{{client_id_key_name }}`, `SOCIAL_AUTH_APPLE_CLIENT`
```
missing whitespace after {{

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1354,6 +1358,12 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+class AppleAuthBackend(SocialAuthMixin, AppleIdAuth):
+    sort_order = 10
+    name = "apple"
+    auth_backend_name = "Apple"
```
we fill want `full_name_validated = True` on this backend I think, for a smoother sign up experience

chdinesh1089 replied:
Okay! added that. also it wouldn't be any problem as the user can change the name later in settings.

mateuszmandera commented on static/styles/portico/portico-signin.scss:
```diff
@@ -677,6 +677,13 @@ button.login-social-button {
+    height: 46px;
+    background-size: auto 100%;
+    background-position: 0px 100%;
+}
```
Ah, this reminds me - we will also want the same for the Sign up button, with the button rendering adjustment in `accounts_home.html` (analogical to `login.html`)

chdinesh1089 replied:
Oh, I missed it. Apple has only two buttons. One saying "Continue with Apple" and other "Sign in with Apple". set both login and register with "Sign in with apple".

gnprice replied:
For the "sign *up*" case, let's use the "Continue with ..." button. That's actually a more accurate meaning for that case than "sign in" is, even though the latter has a word in common.

chdinesh1089 commented:
here are SS's for both login,and register pages.
![login](https://user-images.githubusercontent.com/44665669/77258498-02c1a100-6ca1-11ea-9b4f-168a8fb68809.png)
![register](https://user-images.githubusercontent.com/44665669/77258529-2b499b00-6ca1-11ea-82b4-1f36e4e9aa14.png).

put the button at the end to not make the page look ugly because of the other icons left-aligned. Shouldn't be a problem even if it's on the top also. Mention if it needs to be moved to the top.

gnprice commented on docs/development/authentication.md:
```diff
@@ -42,6 +42,25 @@ details worth understanding:
+* Visit https://developer.apple.com/account/resources/,
+  [Create a Sign In Certificate](https://help.apple.com/developer-account/?lang=en#/dev77c875b7e)
+
+* [Create a Services ID](https://help.apple.com/developer-account/?lang=en#/dev1c0e25352). Fill the redirect URI with
```
URL, not URI.

[The standard term is "URL"](https://chat.zulip.org/#narrow/stream/92-learning/topic/URLs/near/794826). Sometimes an existing API says "URI" and it makes sense to match it; but here, all the Apple docs say "URL" too.

This also applies to the prod instructions.

chdinesh1089 replied:
Ohh, Thanks for pointing to czo topic

gnprice commented on zerver/context_processors.py:
```diff
@@ -152,13 +152,29 @@ def login_context(request: HttpRequest) -> Dict[str, Any]:
+                             'it_IT', 'iw_IL', 'ja_JP', 'ko_KR', 'ms_MY',
+                             'pt_PT', 'ro_RO', 'ru_RU', 'sk_SK', 'sv_SE',
+                             'pl_PL', 'th_TH', 'tr_TR', 'uk_UA', 'vi_VI'
+                             'zh_CN', 'zh_HK', 'id_ID', 'pt_BR']
```
Missing comma gives this the wrong meaning -- there's an element `'vi_VIzh_CN'` :stuck_out_tongue:

gnprice commented on zerver/context_processors.py:
```diff
@@ -152,13 +152,29 @@ def login_context(request: HttpRequest) -> Dict[str, Any]:
+                             'it_IT', 'iw_IL', 'ja_JP', 'ko_KR', 'ms_MY',
+                             'pt_PT', 'ro_RO', 'ru_RU', 'sk_SK', 'sv_SE',
+                             'pl_PL', 'th_TH', 'tr_TR', 'uk_UA', 'vi_VI'
+                             'zh_CN', 'zh_HK', 'id_ID', 'pt_BR']
```
This should get a comment saying where it comes from -- in particular a link to https://developer.apple.com/documentation/signinwithapplejs/incorporating_sign_in_with_apple_into_other_platforms .

I'm also kind of confused by how it's mostly, but not quite, in order. How about putting it in the same order it appears in on that docs page? That'll make it easier to scan with the eye to see they match.

chdinesh1089 replied:
> I'm also kind of confused by how it's mostly, but not quite, in order. 

That's because I shuffled just some at end of each line to have equal line length for all the lines.

I've put them in order and removed these values `'en_GB', 'fr_CA', 'es_MX', 'pt_BR', 'zh_HK',zh_TW'` as they have same language code, thinking these are less popular than those left in the list

gnprice commented on zerver/context_processors.py:
```diff
@@ -152,13 +152,29 @@ def login_context(request: HttpRequest) -> Dict[str, Any]:
+            if current_language_code in locale:
+                return locale
+        return 'en_US'
+
```
This function would be great to have a few tests for.

To make that clean, one way is to pull it out as a top-level function, which can take `request.LANGUAGE_CODE` as an argument. Like this:
```py3
def get_apple_locale(django_language_code: str) -> str:
  # ...
```

There are also some changes I'm tempted to suggest to the algorithm. But that will be much easier to play with once the function is extracted and has some tests.

chdinesh1089 replied:
Added one test which contains list of possible language codes and its corresponding locale, then asserts locale equal to the  return value  `get_apple_locale(language_code)`

chdinesh1089 commented on docs/development/authentication.md:
```diff
@@ -42,6 +42,25 @@ details worth understanding:
+
+* In `dev-secrets.conf`, set `social_auth_apple_id_client` to your
+  "Service ID" (eg. com.application.your),
+  `social_auth_apple_id_key` to your "Key Identity" and
```
Looking at this [picture](https://d33wubrfki0l68.cloudfront.net/c56583180a9951cfb2dc091c0f87a68498df50bb/d0fd5/assets-jekyll/blog/sign-in-with-apple/find-your-key-id-860ac5b63469f699f3c389908d934abf4622c87bccca8c1e178a597c7038cc42.png) from https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple, changing that to Key ID.

chdinesh1089 commented:
Made the suggested changes @gnprice, @mateuszmandera. Thanks for your patience here, there has been a lot of confusion with the terminology.

gnprice commented:
Thanks for the revisions! I think this looks good now for the aspects I've looked at. Eager to see this complete and working in production :smile:

chrisbobbe commented:
Have we looked into the "Email relay service" that will let people receive emails even if they choose not to expose their real email address? See "Email Relay Service" at https://developer.apple.com/sign-in-with-apple/get-started/.

chdinesh1089 commented:
People who are self-hosting would configure an outgoing email account following this https://zulip.readthedocs.io/en/latest/production/email.html#email-services

I think they'll have to add the domain of the email account they configured for zulip to send emails. I guess it's worth mentioning about this in instructions for setting up apple auth in `prod_settings_template.py`.

Once we try setting things up locally, everything will be clear to write precise instructions.

chrisbobbe commented:
> People who are self-hosting would configure an outgoing email account following this https://zulip.readthedocs.io/en/latest/production/email.html#email-services

This makes sense, as one necessary step.

> Once we try setting things up locally, everything will be clear to write precise instructions.

Makes sense. 🙂

> I think they'll have to add the domain of the email account they configured for zulip to send emails.

As you've said, we'll know more once we've set things up locally. But just as a prediction, I think it might be necessary for someone who's self-hosting to contact @gnprice, who has admin privileges in the Apple Developer Console, and Greg will add the domain. If I understand correctly, the domain must be added by an account that has control of the mobile app.

In fact, unless I've misunderstood, isn't it only possible to get the "Sign In with Apple" private key and the Services ID (following [these steps](https://github.com/zulip/zulip/pull/14235/files#diff-d25517512810025038d7e9a56c502281R130-R159)) if you're logged into the Apple Developer Console as Greg? From [this doc](https://help.apple.com/developer-account/?lang=en#/dev1c0e25352):

> To configure web authentication, you must create a Services ID and associate your website to an existing primary iOS, macOS, tvOS, or watchOS App ID enabled for Sign in with Apple.

If this is all true, I suppose we'll just hope people don't mind having to go through Greg to set this up.

gnprice commented:
Yes, a server admin who wants to set up "Sign in with Apple" for their server will need their own account on developer.apple.com, and need to use it to set up their own suite of IDs and credentials. Those won't interact with the credentials we use for running zulipchat.com and chat.zulip.org. In this respect auth with Apple will be just like auth with Google, or GitHub, or the other identity providers we support.


> I think they'll have to add the domain of the email account they configured for zulip to send emails. I guess it's worth mentioning about this in instructions for setting up apple auth in `prod_settings_template.py`.

I agree. Glad we've spotted that!


> I'm actually waiting for greg to provide me private key and Services ID or test locally.

Hmm! That is news to me. :slightly_smiling_face:

I'll start a chat thread to discuss the details of what's needed here. Happy to help.

chrisbobbe commented:
> Yes, a server admin who wants to set up "Sign in with Apple" for their server will need their own account on developer.apple.com, and need to use it to set up their own suite of IDs and credentials.

Just to have this written down, if I understood correctly from my call with @gnprice today, this comes with a step that's a bit non-intuitive, but we're hoping it won't be too inconvenient. From the [doc](https://help.apple.com/developer-account/?lang=en#/dev1c0e25352) I linked to ([above](https://github.com/zulip/zulip/pull/14235#issuecomment-609999218)):

> To configure web authentication, you must [...] associate your website to an existing primary iOS, macOS, tvOS, or watchOS App ID enabled for Sign in with Apple.

To fulfill this, part of the server admin's "own suite of IDs and credentials" is the App ID for a dummy app under their own control, which we're hoping can be pretty minimal. One lever Apple might use to make web authentication much more difficult (but not impossible) for self-hosting folks is to require that this dummy app pass an App Store review, which of course means that it can't be trivial (they have a ["Minimum Functionality" requirement](https://developer.apple.com/app-store/review/guidelines/#4.2)).

gnprice commented:
> To fulfill this, part of the server admin's "own suite of IDs and credentials" is the App ID for a dummy app under their own control, which we're hoping can be pretty minimal.

Yep! And in fact as I found out [later today](https://chat.zulip.org/#narrow/stream/3-backend/topic/Apple.20Auth/near/848503) and confirmed by trying it just now (same link, a few messages later), creating an App ID is super simple given an Apple developer account -- you just go to [Certificates, Identifiers & Profiles > Identifiers](https://developer.apple.com/account/resources/identifiers/list/bundleId), hit the "+" button, and make up a name. Then you can check a box to enable that App ID for Sign in with Apple.

Haven't yet tested that that all works end-to-end -- I think @chdinesh1089 is going to do that next, with his dev server -- but hopefully that takes care of it.

mateuszmandera commented on docs/development/authentication.md:
```diff
@@ -42,6 +42,25 @@ details worth understanding:
+  [Create a Sign In Certificate](https://help.apple.com/developer-account/?lang=en#/dev77c875b7e)
+
+* [Create a Services ID](https://help.apple.com/developer-account/?lang=en#/dev1c0e25352). Fill the redirect URL with
+  `http://zulipdev.com:9991/complete/apple` .
```
as we've noticed this actually needs `/` at the end.

mateuszmandera commented on zerver/context_processors.py:
```diff
@@ -159,6 +160,7 @@ def login_context(request: HttpRequest) -> Dict[str, Any]:
         'password_auth_enabled': password_auth_enabled(realm),
         'any_social_backend_enabled': any_social_backend_enabled(realm),
         'two_factor_authentication_enabled': settings.TWO_FACTOR_AUTHENTICATION_ENABLED,
+        'locale': AppleAuthBackend.get_apple_locale(request.LANGUAGE_CODE)
```
wouldn't `apple_locale` be a better name?

mateuszmandera commented:
Found two minor comments to make, looks good to me aside of that! Let's see what pops up in testing

mateuszmandera commented:
The `state` handling with redis doesn't seem correct - I'm going to try to do some work on this in the following days

chdinesh1089 commented:
Okay! Thought that was good. I'm curious what I did wrong, would try to understand my mistakes from your work :)
 Is it not right to do `backend.get_request_state` in` redis_data = get_dict_from_redis(redis_client, "apple_key_{token}", backend.get_request_state())` ?
Or is it something else?

mateuszmandera commented:
You can look at the SAML implementation - this will be something similar.
Though more difficult, because we'll need to dump the data into redis
somewhere we have access to the `state` as well as the request data - I
don't yet know where the correct place for that will be. SAML puts the data
under a randomly generated key in redis, and then sends that key in the
RelayState to the IdP server - and we get it back in the RelayState in POST
when the user has authed with the IdP and is sent back to us. With Apple,
we'll have to mess with the internals a bit more - to tie the `state` into
this.
That's the high-level description, we'll see the details from the mess that
I'll create trying to implement this 😬

On Sat, Apr 11, 2020 at 8:28 PM Dinesh Ch <notifications@github.com> wrote:

> Okay! Thought that was good. I'm curious what I did wrong, would try to
> understand my mistakes from your work :)
> Is it not right to do backend.get_request_state in redis_data =
> get_dict_from_redis(redis_client, "apple_key_{token}",
> backend.get_request_state()) ?
> Or is it something else?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/14235#issuecomment-612484264>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKXMCMHC7BJ3OLTBZTS3PGTRMCZF7ANCNFSM4LON556Q>
> .
>

chdinesh1089 commented:
fixed the conflict and changed `locale` to `apple_locale` with the force push.

gnprice commented on docs/development/authentication.md:
```diff
@@ -42,6 +42,25 @@ details worth understanding:
+* Visit https://developer.apple.com/account/resources/,
+  [Create a Sign In Certificate](https://help.apple.com/developer-account/?lang=en#/dev77c875b7e)
+
+* [Create a Services ID](https://help.apple.com/developer-account/?lang=en#/dev1c0e25352). Fill the redirect URL with
+  `http://zulipdev.com:9991/complete/apple/` .
```
This comes before creating the private key.

Here's a handy banner Apple shows when you do any one of these steps (as they apparently realize that the process is too complicated and underdocumented):
![image](https://user-images.githubusercontent.com/28173/79810647-62ef4580-8328-11ea-8c6f-04fb343df40e.png)

gnprice replied:
```suggestion
* [Create a Services ID](https://help.apple.com/developer-account/?lang=en#/dev1c0e25352). When prompted for a "Return URL", enter
  `http://zulipdev.com:9991/complete/apple/` .
```

Based on the text in their UI here:
![image](https://user-images.githubusercontent.com/28173/79810838-e3ae4180-8328-11ea-9bc7-17239138a502.png)

chdinesh1089 replied:
made the changes. do we want to make "Services ID", "Service ID"? In the apple banner screenshot they've written "Service ID"

gnprice replied:
Well spotted! :smile: 

They seem to be not quite consistent about it. The docs we're linking to all say "Services ID" with the plural, though -- or at least all the places I remember spotting did. They also say it that way in the part of the UI where you actually go to create one:
![image](https://user-images.githubusercontent.com/28173/79915773-6e03ad80-83dc-11ea-98de-ec2dc538be00.png)
So I think that's the version to stick to, for the least possible amount of inconsistency.

chdinesh1089 replied:
okay, thanks for clarifying. I really like that you share screenshots for most things to make it clear.

gnprice commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,47 @@
     # 'zproject.backends.ZulipRemoteUserBackend',  # Local SSO, setup docs on readthedocs
 )  # type: Tuple[str, ...]
 
+########
+# Apple OAuth.
+#
+# To set up Sign In with Apple, you'll need to do the following:
```
(Same comments as above, modified as appropriate.)

chdinesh1089 commented:
Made changes to Redis implementation after some discussion with @mateuszmandera. This needs to be modified later of "native" apple auth flow. It should now work for browser flow on mobile like other authentication methods.

There's one question I have: 
Been seeing this warning everytime I do manual tesing. Is it a normal thing, or something that needs to be fixed. I don't remember seeing this for GitLab authentication.
![apple](https://user-images.githubusercontent.com/44665669/80379161-56846300-88bb-11ea-9d67-fea90110b635.png)

mateuszmandera commented:
> Been seeing this warning everytime I do manual testing. Is it a normal thing, or something that needs to be fixed. I don't remember seeing this for GitLab authentication.
> ![apple](https://user-images.githubusercontent.com/44665669/80379161-56846300-88bb-11ea-9d67-fea90110b635.png)

I think this has to be because we run the dev environment on http instead of https? In that case, it won't be a problem on production deployments, so nothing to worry about it.

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1364,6 +1373,74 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+        data_to_store = {
+            key: request_data[key] for key in SOCIAL_AUTH_FIELDS_STORED_IN_SESSION if key in request_data
+        }
+        state = put_dict_in_redis(redis_client, 'apple_auth_{token}',
+                                  data_to_store, self.REDIS_EXPIRATION_SECONDS,
+                                  token = self.state_token())
+        state = state.replace('apple_auth_', '')
+        return state
```
```suggestion
        state = self.state_token()
        put_dict_in_redis(redis_client, 'apple_auth_{token}',
                          data_to_store, self.REDIS_EXPIRATION_SECONDS,
                          token=state)

        return state
```

that's simpler i think?

chdinesh1089 replied:
yes, it's much simpler thanks!

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1364,6 +1373,74 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+        if not request_state:
+            raise AuthMissingParameter(self, 'state')
+        elif redis_data is None:
+            raise AuthStateForbidden(self)
```
why this exception specifically?

chdinesh1089 replied:
the original `validate_state` function had that when both states do not match. So, It is similar here if we do not get redis_data as it means we get a wrong state value.

mateuszmandera replied:
Makes sense!

timabbott replied:
Should we be calling `validate_state` from the superclass rather than reimplementing parts of it?

mateuszmandera replied:
Not really, we're re-doing the whole method. The upstream one compares the `state` in the request with the `state` saved in user's session. But Apple sends the user here with a POST request, so we don't have the session.

mateuszmandera commented on zproject/backends.py:
```diff
@@ -63,7 +65,7 @@
     get_user_profile_by_id, remote_user_to_email, \
     email_to_username, get_realm, get_user_by_delivery_email, supported_auth_backends
-
+from zproject.settings import SOCIAL_AUTH_FIELDS_STORED_IN_SESSION
```
I think we should avoid this import and just use `settings.SOCIAL_...` the way we always do :thinking:

chdinesh1089 replied:
changed.

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1445,6 +1453,79 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+        state = self.state_token()
+        put_dict_in_redis(redis_client, 'apple_auth_{token}',
+                          data_to_store, self.REDIS_EXPIRATION_SECONDS,
+                          token = state)
```
`token=state`, we don't do the whitespaces when specifying kwargs, as a convention

chdinesh1089 replied:
Didn't knew. Thanks! changed it.

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1445,6 +1453,79 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+            logging.info("Sign in with Apple failed: missing state parameter.")
+            raise AuthMissingParameter(self, 'state')
+
+        formatted_request_state = "apple_auth_"+request_state
```
```suggestion
        formatted_request_state = "apple_auth_" + request_state
```

(whitespaces around arithmetic operations as a convention)

mateuszmandera commented on zproject/backends.py:
```diff
@@ -1445,6 +1453,79 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+        if redis_data is None:
+            logging.info("Sign in with Apple failed: bad state token.")
+            raise AuthStateForbidden(self)
+        else:
```
This will be more readable without the `else` and it's redundant as the `if` above `raises` anyway

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1092,6 +1096,7 @@ def stage_two_of_registration(self, result: HttpResponse, realm: Realm, subdomai
             self.verify_desktop_flow_end_page(result, email, desktop_flow_otp)
             # Now the desktop app is logged in, continue with the logged in check.
         else:
+            print(result)
```
debugprint got lost here :slightly_smiling_face:

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1661,6 +1666,143 @@ def test_social_auth_saml_require_limit_to_subdomains(self) -> None:
+        # return 'en_US' if invalid `language_code` is given.
+        self.assertEqual(AppleAuthBackend.get_apple_locale(':)'), 'en_US')
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
```
we should put this function definition above the test cases, like in other classes

chdinesh1089 replied:
changed

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1661,6 +1666,143 @@ def test_social_auth_saml_require_limit_to_subdomains(self) -> None:
+        token_data_dict = {
+            'access_token': 'foobar',
+            'expires_in': time.time() + 60*5,
+            'id_token': 'blah blah',
```
This + the `decode_id_token` mock do the job, though it would be nicer if we managed to actually make a valid `id_token` here. If it's difficult, then probably not worth it, but look at how the PSA apple backend implementation creates some data:
```
    def generate_client_secret(self):
        now = int(time.time())
        client_id = self.setting('CLIENT')
        team_id = self.setting('TEAM')
        key_id = self.setting('KEY')
        private_key = self.get_private_key()

        headers = {'kid': key_id}
        payload = {
            'iss': team_id,
            'iat': now,
            'exp': now + self.TOKEN_TTL_SEC,
            'aud': self.TOKEN_AUDIENCE,
            'sub': client_id,
        }

        return jwt.encode(payload, key=private_key, algorithm='ES256',
                          headers=headers)
```

perhaps we can use the same method to generate `id_token` with the same data structure that the real Apple-generated `id_token` use? We have the private key in test fixtures already.

chdinesh1089 replied:
the actual `decode_id_token` gets public key from https://appleid.apple.com/auth/keys and decodes `id_token`. so, we'll have to make a different private key(we could use the one we have too), encode it, use the public key that is pair of private key.

The reason behind mocking it was that I was thinking testing `decode_id_token` is already done by PSA and we wouldn't need it.

Shall I try doing it? We would probably have to add some extra code there to encode `account_data_dict` to get `id_token` there

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1661,6 +1666,143 @@ def test_social_auth_saml_require_limit_to_subdomains(self) -> None:
+        # The user is taken to registration form, where the following enters
+        # "Full Name" in that field.
+        self.stage_two_of_registration(result, realm, subdomain, email, '',
+                                       'Full Name', skip_registration_form = False)
```
```suggestion
                                       'Full Name', skip_registration_form=False)
```

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1066,7 +1066,11 @@ def stage_two_of_registration(self, result: HttpResponse, realm: Realm, subdomai
             self.assert_in_success_response(['id_full_name'], result)
             # Verify the name field gets correctly pre-populated:
-            self.assert_in_success_response([expected_final_name], result)
+            if (self.BACKEND_CLASS != AppleAuthBackend):
```
Makes sense, though unfortunate we have to do that :cry: What we can do to avoid polluting this helper function with this apple-specific logic, I think we can do a different trick and override `stage_two_of_registration` in the apple backend to simply call `super().stage_two_of_registration(...)` with `expected_final_name=''`? Same effect, though more hacky - but we manage to keep this helper cleaner at least, so might be worth it

chdinesh1089 replied:
I'm not sure if I understood the right way about override, you'e saying that we'll just make the content of the overriden function this?
``` 
super().stage_two_of_registration(......,expected_final_name='')
```
If we do that it'd fail the test because we cannot have full name here to be empty string I guess.
```
# Click confirm registration button.
            result = self.client_post(
                '/accounts/register/',
                {'full_name': expected_final_name,
                 'key': confirmation_key,
                 'terms': True})
```
Now I see that that change is only for one test `test_auth_registration_with_no_name_sent_from_apple`. so, I'll copy the contents of `stage_two_of registration` to only have relevant parts.

mateuszmandera commented:
Looks great so far, I'm super happy we're getting so close to having this ready :slightly_smiling_face:

chdinesh1089 commented:
Made the suggested changes except the `id_token` one. I'd try that too if we need that. But I actually thought PSA does the required testing for that particular `decode_id_token` function.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1661,6 +1661,203 @@ def test_social_auth_saml_require_limit_to_subdomains(self) -> None:
+            # `id_token` we have the right public key to decode it.
+            # This mock return's that public key in jwk format.
+            with mock.patch('social_core.backends.apple.AppleIdAuth.get_apple_jwk',
+                            return_value=settings.APPLE_JWK):
```
Could this be replaced with a sample RequestsMock of the apple url to return the JWK as http response?

chdinesh1089 replied:
yes, replaced it.

mateuszmandera commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,47 @@
+# Follow the instructions in
+# https://help.apple.com/developer-account/?lang=en#/dev1c0e25352.
+# to Create a Services ID. In services ID, When prompted for a "Return URL",
+# enter `http://zulip.example.com/complete/apple/`.
```
```suggestion
# enter `https://zulip.example.com/complete/apple/`.
```

mateuszmandera commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,47 @@
+# https://help.apple.com/developer-account/?lang=en#/dev77c875b7e
+# to create a Sign In with Apple private key.
+#
+# (3) You'll now have a "Services ID"(eg: com.application.your),a "Key ID"
```
```suggestion
# (3) You'll now have a "Services ID" (eg: com.application.your), a "Key ID"
```

chdinesh1089 commented:
Made the suggested changes. Changed the path to save apple private key to `zproject/dev_apple.key` and changed instructions in `authentication.md` saying store the private key in path `zproject/dev_apple.key`.

Also added the instruction to set `team_id` which I missed earlier for dev setup instructions. sorry for the trouble cause @chrisbobbe because of missing this part.

@mateuszmandera do we need to update `.gitignore` to ignore `zproject/dev_apple.key` and `zproject/dev_saml.cert` to avoid accidental commits?

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1891,6 +1891,65 @@ def test_validate_state(self) -> None:
             m.assert_called_with("Sign in with Apple failed: bad state token.")
 
+class AppleAuthBackendNativeFlowTest(AppleIdAuthBackendTest):
+    SIGNUP_URL = '/complete/apple/'
```
I think we need to do LOGIN_URL as well?

mateuszmandera replied:
I'm a bit confused about why this is even working without overriding LOGIN_URL

chdinesh1089 replied:
It works because we set ` alternative_start_url=self.AUTH_FINISH_URL` in `prepare_login_url_and_headers` call.
we cannot use it for signup case because even if we give an `alternative_start_url` it's gets  changed to `SIGNUP_URL`. (I guess it makes sense to have it like that for other backends)

chdinesh1089 replied:
I think we do not need to change `LOGIN_URL` and we will not be having a `LOGIN_URL` for native test class as we do not include that in the mixin. The tests should work perfectly as we do not go to `LOGIN_URL` at any point in native flow afaik.

chdinesh1089 replied:
There's this test `test_social_auth_session_fields_cleared_correctly` that uses actual `prepare_login_url_and_headers` which uses actual `LOGIN_URL` and doesn't make use of `social_auth_test`. 
If we're using the actual `LOGIN_URL` there's no difference with running this test in `AppleIdAuthBackendTest` and `AppleAuthBackendNativeFlowTest`.

I added a test overriding this doing it right way for this flow.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1891,6 +1891,65 @@ def test_validate_state(self) -> None:
             self.assertIn('login', result.url)
             m.assert_called_with("Sign in with Apple failed: bad state token.")
 
+class AppleAuthBackendNativeFlowTest(AppleIdAuthBackendTest):
```
This should subclass SocialAuthBase because `AppleIdAuthBackendTest` has some tests that don't make sense to run in the native flow context

mateuszmandera replied:
The helper functions from AppleIdAuthBackendTest that we need here can be extracted into a mixin class.

chdinesh1089 replied:
created a mixin class `AppleAuthMixin`

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1891,6 +1891,65 @@ def test_validate_state(self) -> None:
+        if subdomain:
+            params['subdomain'] = subdomain
+
+        url += "&%s" % (urllib.parse.urlencode(params),)
```
I think we should just override `prepare_login_url_and_headers` to do `super().prepare_login_url_and_headers` and then this logic above - because that's where the url construction naturally belongs.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1891,6 +1891,65 @@ def test_validate_state(self) -> None:
+        pass
+
+    def test_social_auth_no_key(self) -> None:
+        pass
```
Can you add comments on these two disabled tests explaining why we're doing that for the native flow?

chdinesh1089 replied:
added.

timabbott replied:
I don't understand why that means we can't / shouldn't run these tests; can you explain?

chdinesh1089 replied:
These tests get the `result` from the following `if` in `social_auth_test` and do not reach the end where `AUTH_FINISH_URL` is called
```
if result.status_code != 302 or not result.url.startswith(expected_result_url_prefix):
            return result
```
Since we skip all of that code here in the native flow test, we cannot expect those tests to work.

chdinesh1089 commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1891,6 +1893,97 @@ def test_validate_state(self) -> None:
+    '''
+    This class is to test the native flow we continue with the
+    data sent as GET request from mobile app.
+    '''
```
Tests coverage isn’t there because __unittest_skip = False isn’t set here. Will fix it.

chdinesh1089 replied:
Fixed it! could you review again @mateuszmandera :slightly_smiling_face:

timabbott commented on zproject/backends.py:
```diff
@@ -1328,6 +1328,8 @@ class SocialAuthMixin(ZulipAuthMixin, ExternalAuthMethod):
     # it should be False.
     full_name_validated = False
 
+    standard_relay_params = settings.SOCIAL_AUTH_FIELDS_STORED_IN_SESSION + ['next']
```
@mateuszmandera should we just be moving `next` into the setting and using the setting here?

mateuszmandera replied:
We have our own plumbing of the `next` parameter throughout the auth process, so adding it to `SOCIAL_AUTH_FIELDS_STORED_IN_SESSION` seems error prone. Would be nice to clean it up if possible though, as a follow-up perhaps

timabbott commented:
I merged the first commit, since it looked good and it'll shrink this branch a bit.

mateuszmandera commented on docs/development/authentication.md:
```diff
@@ -42,6 +42,23 @@ details worth understanding:
+* [Create a Sign in with Apple private key](https://help.apple.com/developer-account/?lang=en#/dev77c875b7e)
+
+* In `dev-secrets.conf`, set `social_auth_apple_client` to your
+  "Services ID" (eg. com.application.your),
```
Shouldn't this be `App ID`, not `Services ID`? At least that's what seems right based on @chrisbobbe testing

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -120,12 +120,49 @@
+#
+#SOCIAL_AUTH_APPLE_SERVICES_ID = <your Services ID>
+#SOCIAL_AUTH_APPLE_BUNDLE_ID = <your Bundle ID>
+#SOCIAL_AUTH_APPLE_TEAM = <your Team ID>
```
A few requests:
* Let's move most of this documentation to `docs/production/authentication-methods.md`; that's HTML, so we can do links and things, and is easier to update (`prod_settings_template.py` gets copied at installation and never updated).
* Let's order the comment block for Apple auth just before `AzureAD` in this file; it's not something individual self-hosted installations are likely to use often.  Google OAuth, the one the most people configure, should still be first.

chdinesh1089 replied:
Moved and reordered. Added quotes for those.
(I've seen that other's don't have these quotes. Do we want to change it for them too?)

timabbott commented on zproject/urls.py:
```diff
@@ -574,7 +574,6 @@
     url(r'^config-error/(?P<error_category_name>[\w,-]+)$', zerver.views.auth.config_error_view,
         name='config_error'),
     url(r'^config-error/remoteuser/(?P<error_category_name>[\w,-]+)$', zerver.views.auth.config_error_view)
-
```
Let's remove this change from the PR.

chdinesh1089 replied:
Added that line back.

timabbott commented on zproject/test_settings.py:
```diff
@@ -173,6 +173,15 @@ def set_loglevel(logger_name: str, level: str) -> None:
+SOCIAL_AUTH_APPLE_TEAM = 'TEAMSTRING'
+SOCIAL_AUTH_APPLE_SECRET = get_from_file_if_exists("zerver/tests/fixtures/apple/private_key.pem")
+
+APPLE_JWK = get_from_file_if_exists("zerver/tests/fixtures/apple/jwk")
```
We should fix `zerver/tests/fixtures/apple/jwk` not having a newline at end-of-file.

chdinesh1089 replied:
Fixed.

timabbott commented on zproject/test_settings.py:
```diff
@@ -173,6 +173,15 @@ def set_loglevel(logger_name: str, level: str) -> None:
+SOCIAL_AUTH_APPLE_CLIENT = 'com.zulip.chat'
+SOCIAL_AUTH_APPLE_KEY = 'KEYISKEY'
+SOCIAL_AUTH_APPLE_TEAM = 'TEAMSTRING'
+SOCIAL_AUTH_APPLE_SECRET = get_from_file_if_exists("zerver/tests/fixtures/apple/private_key.pem")
```
Can we add a comment, ideally in the file itself, but worst case here, making clear how this private key was generated and why it's safe to have it checked into version control?

chdinesh1089 replied:
Added a comment in the file with the command used to generate the file, the reason we needed it and saying it's safe to have it checked into version control.

timabbott commented on zproject/backends.py:
```diff
@@ -1488,8 +1628,6 @@ def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
 class SAMLAuthBackend(SocialAuthMixin, SAMLAuth):
     auth_backend_name = "SAML"
-    standard_relay_params = ["subdomain", "multiuse_object_key", "mobile_flow_otp", "desktop_flow_otp",
-                             "next", "is_signup"]
```
Can you rebase?  This hunk will disappear from the diff as we've already merged this commit.

timabbott commented on static/styles/portico/portico-signin.scss:
```diff
@@ -682,6 +682,15 @@ button.login-social-button {
 }
 
+button#login_auth_button_apple,
+button#register_auth_button_apple {
```
Let's use CSS classes rather than IDs for these.  

Also, please add a comment explaining why this requires different geometry from our default social buttons.

chdinesh1089 replied:
The IDs were generated like this `<button id="register_{{ backend.button_id_suffix }}" class="login-social-button full-width"`. Using IDs lets us require fewer changes in the HTML pages. Is it okay to add a conditional for that to have a class for this particular one?

(Added a comment. Also changed the `height: 46px` to 50px to make those have the same height as other buttons.)

timabbott replied:
Yeah, I guess this is OK.

timabbott commented on templates/zerver/accounts_home.html:
```diff
@@ -78,7 +78,8 @@ <h1 class="get-started">{{ _("Sign up for Zulip") }}</h1>
                                 <input type='hidden' name='multiuse_object_key' value='{{ multiuse_object_key }}' />
                                 <button id="register_{{ backend.button_id_suffix }}" class="login-social-button full-width"
-                                  {% if backend.display_icon %} style="background-image:url({{ backend.display_icon }})"  {% endif %}>
+                                  {% if backend.display_name == "Apple" %} style="background-image: url(https://appleid.cdn-apple.com/appleid/button/?width=328&height=46&color=white&locale={{ apple_locale }}&type=continue);"
```
Why can't this computation be done in the called and used to populate `backend.display_icon`?

mateuszmandera replied:
We need to use different button url on login and signup page, plus the `{{ apple_locale }}` thing needs to be here and it's generated based on the request (based on language of the browser. Doesn't seem to be any good way to have this code anywhere in the backend class, and of all other places to put this Apple-specific code this is probably the least awful one. In an ideal world Apple would restrain themselves from forcing absurd rules on people at least when it comes to buttons, and we would be to use our regular buttons with a simple Apple logo...

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1872,6 +1872,364 @@ def test_idp_initiated_signin_subdomain_implicit_invalid(self) -> None:
+    # failing because of assigning `url=self.LOGIN_URL` which gets
+    # changed because of `alternative_start_url` we send as a param
+    # and has no role in any test.
+    LOGIN_URL = ''
```
I don't understand this comment at all.

chdinesh1089 replied:
seeing this code makes it clear  I think.
```
def prepare_login_url_and_headers(
        self,
        subdomain: Optional[str]=None,
        mobile_flow_otp: Optional[str]=None,
        desktop_flow_otp: Optional[str]=None,
        is_signup: bool=False,
        next: str='',
        multiuse_object_key: str='',
        alternative_start_url: Optional[str]=None,
        *,
        user_agent: Optional[str]=None,
    ) -> Tuple[str, Dict[str, Any]]:
        url = self.LOGIN_URL
        if alternative_start_url is not None:
            url = alternative_start_url

        params = {}
        headers = {}
        if subdomain is not None:
....
...
`
there's `url = self.LOGIN_URL` before url gets assigned to `alternative_start_url` in the `if` block. If `LOGIN_URL` is missed, that functions fails since it cannot find `LOGIN_URL`. 
As url is getting changed every time to `alternative_url` replacing `LOGIN_URL`, we say `LOGIN_URL` has no role in tests.
I'll try to make it a bit clear.

chdinesh1089 replied:
Modified the comment. Hope it's clear now :slightly_smiling_face:

timabbott commented on zproject/backends.py:
```diff
@@ -70,7 +75,6 @@
     EmailContainsPlusError, PreregistrationUser, UserProfile, Realm, custom_profile_fields_for_realm, \
     get_user_profile_by_id, remote_user_to_email, \
     email_to_username, get_realm, get_user_by_delivery_email, supported_auth_backends
-
```
Add this newline back; it was intentional.

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+    sort_order = 10
+    name = "apple"
+    auth_backend_name = "Apple"
+    full_name_validated = True
```
This probably deserves a comment that full_name isn't available on repeat logins, but when it is, it is validated.

chdinesh1089 replied:
Added.

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+                logging.info("/complete/apple/: %s", str(e))
+                return None
+        else:
+            return super().auth_complete(*args, **kwargs)
```
Can you use the early return pattern to handle this case first, and deindent the rest of the function?  I think that'd be more readable.  We should also add a comment explaining that the `python-social-auth` default implementation is the web flow, so we can just delegate to them for that; and we just need custom code for the native flow.

chdinesh1089 replied:
Removed the `else` and deindented. Added a comment explaing we use `auth_complete` of `python-social-auth` when it's not native flow.

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+            key: request_data[key] for key in self.standard_relay_params
+            if key in request_data
+        }
+        state = self.state_token()
```
Can you add a comment here explaining what format `state_token` has and what assumptions we can make about it (E.g. is it a unique random string generated by us, or is it under Apple or the user's control?)

chdinesh1089 replied:
Added a comment. 
(It's a random string generated using PSA's internal function `state_token`)

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
     # (No code is required to do so, as that's the default behavior).
 
+@external_auth_method
+class AppleAuthBackend(SocialAuthMixin, AppleIdAuth):
```
Can you add a block comment here explaining the basic flow of `AppleIdAuth`?  I'm not particularly familiar with JWT, and in any case we need this to be readable be someone working on Zulip's authentication code who isn't even if I am.

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+        return 'en_US'
+
+    def get_or_create_state(self) -> str:
+        '''
```
Please clarify clearly any methods from the upstream backend that we're overriding.

chdinesh1089 replied:
Added a comment.
Also added another comment above that saying all the functions below are the function that we had to override.

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+            decoded = jwt.decode(id_token, key=public_key,
+                                 audience=audience, algorithm="RS256")
+        except PyJWTError:
+            raise AuthFailed(self, "Token validation failed")
```
What is the difference between `AuthCanceled` and `AuthFailed`?  That seems to be a change we made here.

For functions like this that we've forked just one line of, we should have a clear comment explaining what we changed and why.  (We may want to submit upstream PRs for those changes to help clean up our implementation; e.g. I would think we'd try to submit upstream a change for this one to accept an optional `audience` argument).

mateuszmandera replied:
From upstream:
```
class AuthCanceled(AuthException):
    """Auth process was canceled by user."""
```
```
class AuthFailed(AuthException):
    """Auth process failed for some reason."""
```

Clearly should be Authfailed here. On top of that, upstream does `raise AuthCanceled("Token validation failed")` which is wrong, because the first argument is the backend, the second arg is the message. Upstream exceptions from here convert to the empty string when doing `str(e)` because of that. I didn't submit a PR upstream, because the maintainer was gone for a couple of months again - but seems to have returned recently, so we should submit these fixes.

They also merged this https://github.com/python-social-auth/social-core/commit/7d3daf501a16cb152298617fa73ecbbf4e129414 which allows passing an `AUDIENCE` list of allowed audiences. But I'm not sure if we like that, because we discussed with Greg a bit in https://chat.zulip.org/#narrow/stream/107-kandra/topic/Apple.20auth/near/886762 and at that point decided to be more strict in the validation - BUNDLE_ID in native flow, SERVICE_id in browser flow. So we might want to keep our custom code here anyway...

timabbott replied:
Or submit a PR upstream to make it better.

mateuszmandera replied:
Yeah. I submitted the error handling fixes for now at https://github.com/python-social-auth/social-core/pull/476

timabbott replied:
OK.  Maybe worth also adding a commit to make `audience` an option argument, which would let us more fully unfork this?

timabbott commented on zproject/backends.py:
```diff
@@ -1470,6 +1483,133 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+                # AuthFailed is a general "failure" exception from PSA that we should convert
+                # to None return value here to avoid getting tracebacks.
+                # AuthCanceled is raised in the Apple backend implementation in PSA
+                # in certain cases, though AuthFailed would have been more correct.
```
Why is AuthFailed more correct than AuthCanceled?  The comment may not need to change, but I'd like to have a clear understanding of why this comment is accurate, since it seems like an important detail.

mateuszmandera replied:
From upstream:
```
class AuthCanceled(AuthException):
    """Auth process was canceled by user."""
```
```
class AuthFailed(AuthException):
    """Auth process failed for some reason."""
```

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1872,6 +1872,364 @@ def test_idp_initiated_signin_subdomain_implicit_invalid(self) -> None:
+                status=200,
+                json=json.loads(settings.APPLE_JWK),
+            )
+            result = self.client_get(url, **headers)
```
Why is this here, rather than in `prepare_login_url_and_headers`?  I'm also a little puzzled why this seems to not actually call the upstream `social_auth_test` method -- if intentional, it certainly deserves a big comment, because it means we're not running most of our test code.

chdinesh1089 replied:
Because we want to test the native flow code, which is just sending `GET` request with `native_flow=true` and the other parts of upstream `social_auth_test` are already tested, we start the flow directly with `result = self.client_get(url, **headers)`. We have it there instead of `prepare_login_url_headers` as we need to mock the request requiring internet that this auth flow makes to get apple public JWK from `https://appleid.apple.com/auth/keys.`.

chdinesh1089 replied:
Added a comment explaining a bit about native flow and why we start with `self.client_get(url, **headers)`.
We can have the remaining code also but we don't really focus on that in testing native flow and it's tested already in `AppleIdAuthBackendTest`.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1872,6 +1872,364 @@ def test_idp_initiated_signin_subdomain_implicit_invalid(self) -> None:
+            self.client.session.flush()
+            result = self.client_post(self.AUTH_FINISH_URL,
+                                      dict(state=state), **headers)
+        return result
```
Do we need to copy all of `social_auth_test`?  That function is designed to be extensible, with e.g. `self.register_extra_endpoints` (which handles at least one of the issues here), and I'd really like to avoid making a bunch of almost-copies of it as that's bad for maintainability.

I think you can achieve what you're trying to do by just using its existing pluggable features and maybe adding a new argument to pass the extra `token_data_dict` data.

chdinesh1089 replied:
That looks to be the best way. made use of `social_auth_test_finish` and `register_extra_endpoins` and removed `social_auth_test`

timabbott commented:
Thanks for working on this @chdinesh1089!  I posted a large batch of comments on this.  I think it'd also be constructive to split this into two commits, one that implements the basic Apple web flow, and a second doing the native flow.  I **think** the commit split would be pretty simple in that it'd mostly be skipping blocks of code that are only needed for native), and that first web-flow commit would be a relatively straightforward merge now; most of the code here that needs work is in the native flow extension (since it involves most of the messy subclassing).  

If you're not comfortable doing that @chdinesh1089, it could make sense for you to fix all the issues I mentioned above, and then hand this off to @mateuszmandera to do that split.

chdinesh1089 commented:
Made the most of the suggested changes and pushed as a different commit so that it's easier for @mateuszmandera to review and then it could be squashed.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1884,12 +1884,16 @@ class AppleAuthMixin:
         # with `SOCIAL_AUTH_APPLE_CLIENT` as the `audience`
         payload['aud'] = settings.SOCIAL_AUTH_APPLE_CLIENT
+
+        if aud is not None:
```
What is `aud`?  We should expand it if it's an abbreviation.

timabbott replied:
Oh, `audience`; yeah, let's make the parameter to the function `audience`, at least, even if we need to put it into a dict as `aud`.  More readable that way.

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -2040,6 +2044,166 @@ def test_validate_state(self) -> None:
+                         multiuse_object_key: str='',
+                         alternative_start_url: Optional[str]=None,
+                         user_agent: Optional[str]=None,
+                         **extra_data: Any) -> HttpResponse:
```
Do we need to do our own `social_auth_test`?  It was really nice to avoid that with the non-native version, and I'd like to do so here as well if we can.

mateuszmandera replied:
I think we really do need our own. Normal social_auth_test is designed for Oauth-like flows. Our native flow has nothing to do with Oauth, it is super straightforward - the app sends us `native_flow=true&id_token=<what the app got from Apple>` to `/complete/apple/` and that's the end of the story. Normal social_auth_test cannot be adapted to something so simple :slightly_smiling_face: 

Which might raise the question "should we even be subclassing SocialAuthBase in this case?". I think the answer is yes, because while the flow is different, we just do some basic things in auth_complete and then hand things over back to PSA, and then they go through social pipeline, where most of our magic happens anyway, all the validation, mobile flow handlings etc. So it's important to test that we did plays well with the social pipeline - by running the SocialAuthBase test suite.

timabbott commented on zproject/backends.py:
```diff
@@ -1655,6 +1661,33 @@ def decode_id_token(self, id_token: str) -> Dict[str, Any]:
+            # The default implementation in python-social-auth is the browser flow.
+            return super().auth_complete(*args, **kwargs)
+
+        # We handle the native flow on our own.
```
Can you expand this to give more discussion of the high-level strategy.  I see we go towards calling `do_auth`, like the original python-social-auth code does ... what is the intent of our changes from the base logic here?

mateuszmandera replied:
Posted a larger block comment detailing things

timabbott commented on zproject/backends.py:
```diff
@@ -1655,6 +1661,33 @@ def decode_id_token(self, id_token: str) -> Dict[str, Any]:
+            result = self.do_auth(request_data['id_token'], *args, **kwargs)
+            return result
+        except (AuthFailed, AuthCanceled) as e:
+            # AuthFailed is a general "failure" exception from PSA that we should convert
```
spell python-social-auth; abbreviations make for hard to understand code.

mateuszmandera commented:
I'll make some remaining fixes

timabbott commented:
Merged as a10f3893711e423e4a5e77e1e03db316c9257904, after a bit more work on the comments/documentation.  Thanks @chdinesh1089 and @mateuszmandera!  

Can both of you do a bit of an audit for where we mention GitLab in our documentation to see if there's anywhere that should be updated to list Apple as well?  Thanks!

chdinesh1089 commented:
We might want to include it in templates/zerver/features.html, templates/zerver/help/configure-authentication-methods.md, templates/zerver/help/logging-in.md

Opened https://github.com/zulip/zulip/pull/15386 including apple in those files

Do we want to include that in https://zulip.readthedocs.io/en/latest/production/authentication-methods.html?highlight=sso#plug-and-play-sso-google-github-gitlab ? I'm not sure because we have [Sign in with Apple](https://zulip.readthedocs.io/en/latest/production/authentication-methods.html?highlight=sso#sign-in-with-apple) subsection in the same page
